### ISSUE TITLE
Support prereleases

### ISSUE THREAD

**Issue by jakobrosenberg** [felipecrs/semantic-release-vsce#192]
I know VSC plugins don't support prereleases, but since it's recommended to use Github releases to handle regression errors, we could also use Github to handle beta releases.

It's very possible that I'm missing an obvious solution to this. 😅

**Comment by felipecrs** [comment]
Usually, what extensions do to handle this is to have a _Preview_ variant of it. Anyway, I thought on 2 things that needs to be made here:

1. Allow semantic-release-vsce to not publish anything, but just package (so that you can call semantic-release-github to create a GitHub release with the VSIX)
2. Allow semantic-release-vsce to specify custom extension ID depending on the branch (the preview one). I'm not sure yet how it would work.

**Comment by felipecrs** [comment]
@jakobrosenberg in fact now pre-releases are supported by `vsce`: https://code.visualstudio.com/updates/v1_63#_pre-release-extensions

**Comment by felipecrs** [comment]
But some work is still required at `vsce` (and maybe in Visual Studio Marketplace as well). I created an issue there to track: https://github.com/microsoft/vscode-vsce/issues/688

**Comment by jakobrosenberg** [comment]
@felipecrs that's pretty cool! Thanks for the heads up. 😀

**Comment by anacierdem** [comment]
I think this is in VSCE for some time now: https://code.visualstudio.com/api/working-with-extensions/publishing-extension#prerelease-extensions

**Comment by felipecrs** [comment]
Unfortunately not. Their implementation is still misaligned with semver which is what we need. I suggest you weigh in https://github.com/microsoft/vsmarketplace/issues/310.

**Comment by anacierdem** [comment]
Ah, I misread the issue, I thought this was about providing the `--pre-release` flag to vsce. I think this is already possible via `package.json`

**Comment by JoseVSeb** [comment]
Regardless of semver not being supported by vs marketplace, can I at least pass `--pre-release` flag for `publish` command through this plugin? If that were possible, I could use logic in semantic release config to set a version as pre release.

**Comment by felipecrs** [comment]
Can you elaborate? What logic would that be?

**Comment by JoseVSeb** [comment]
in release config, i could check the branch to pass the prerelease flag to this plugin, something like this:
```
// release.config.js
export default {
  branches: ["main", "next"],
  plugins: [
    [
      'semantic-release-vsce',
      {
        prerelease: GIT_BRANCH === "next",
      },
    ],
  ],
};
```

**Comment by felipecrs** [comment]
Ok, but then what would semantic-release-vsce do with this flag?

**Comment by felipecrs** [comment]
Sorry, I misread your first comment. Let me think about it.

**Comment by JoseVSeb** [comment]
forward it to `vsce publish --pre-release` as shown [here](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#prerelease-extensions)

**Comment by felipecrs** [comment]
Ok, that can actually work.

In this case, from semantic-release's perspective this version would be a _release_ version (not a _pre-release_) that would not be published to the _latest_ channel.

The GitHub Release will not be marked as _pre-release_, but it will also not be marked as _latest_.

And the VS Marketplace release would be marked as _pre-release_.

---

But introducing a `prerelease` flag for `semantic-release-vsce` would be confusing. People would think that had something to do with semantic-release's _pre-releases_.

Any ideas to make the flag name less confusing?

**Comment by JoseVSeb** [comment]
Maybe `marketplacePreRelease` or `publishAsPreRelease`? Not sure if any better.

Also, OpenVSX publish should be considered.

**Comment by felipecrs** [comment]
`publishAsPreRelease` seems good. From what I can tell, OpenVSX publish works the same way, so we can simply extend the flag for it too.

I'll accept a contribution for this, or maybe even do it myself. I just recommend waiting for #331 to avoid unnecessary rework. I plan to finish it this weekend.

**Comment by felipecrs** [comment]
@JoseVSeb you can now enable `--pre-release` with `VSCE_PRE_RELEASE=true`:

- https://github.com/felipecrs/semantic-release-vsce/pull/976

Note I didn't test it for real, but it should work. Any feedback is appreciated.

**Comment by JoseVSeb** [comment]
@felipecrs Thanks for the support! 👍
The prerelease works as intended and I've used it [here](https://github.com/JoseVSeb/google-java-format-for-vs-code/releases/tag/v1.2.2)
The GitHub Release is marked as prerelease, along with the publish to VS Marketplace and OpenVSX.

**Comment by JoseVSeb** [comment]
For publish to OpenVSX, it appears the prerelease is also tagged with latest at the same time. 🤔
[![Screenshot_20260308-233931.png](https://github.com/user-attachments/assets/b2f5f8b8-c1b7-4061-a780-d54b647848d1)](https://open-vsx.org/extension/josevseb/google-java-format-for-vs-code)

**Comment by felipecrs** [comment]
> For publish to OpenVSX, it appears the prerelease is also tagged with latest at the same time. 🤔

Yeah. That doesn't seem proper, but that's more likely to be an issue with OpenVSX itself. Maybe you can raise a bug report there?

To mimic what semantic-release-vsce does:

```console
$ vsce package --pre-release ...
$ ovsx publish --packagePath <path to .vsix>
```
